### ISSUE TITLE
IfcMaterial ids lost when transformed from IFC to XML

### ISSUE THREAD

**Issue by bingo-soft** [IfcOpenShell/IfcOpenShell#5109]
Hi there! When I convert my IFC model to XML, I see that some important attributes like "ids" are missing. This is what I get as a result of running IfcConvert:

     
      <IfcMaterialList id="IfcMaterialList_56413">
	    <IfcMaterial Name="Metal"/>
	    <IfcMaterial Name="Metal"/>
	    <IfcMaterial Name="Wood"/>
	    <IfcMaterial Name="Wood"/>
	    <IfcMaterial Name="Wood"/>
      </IfcMaterialList>

Howevr, in IFC itself I see that all these child materials have their "id" attributes. This is the command I used:

`IfcConvert -j 2 model.ifc model.xml -qy`

I also tried options --use-material-names and --use-element-guids, but they do not help. What is wrong with that and how can I get correct XML from my IFC model? Thanks!

**Comment by aothms** [comment]
Are you interested in contributing a fix for this. https://github.com/IfcOpenShell/IfcOpenShell/blob/6cb063ebd2e4f79b1c679b30e31af82d679b5cf5/src/serializers/schema_dependent/XmlSerializer.cpp#L798 More eyes on the code base always helps :)

**Comment by bingo-soft** [comment]
> Are you interested in contributing a fix for this.
> 
> https://github.com/IfcOpenShell/IfcOpenShell/blob/6cb063ebd2e4f79b1c679b30e31af82d679b5cf5/src/serializers/schema_dependent/XmlSerializer.cpp#L798
> More eyes on the code base always helps :)

I have never touched IfcOpenShell code, but it seems like one single line of code will do the trick:

`subnode.put("<xmlattr>.Id", (*jt)->Material()->Id());`

**Comment by aothms** [comment]
I don't see an Id attribute on IfcMaterial https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcMaterial.htm It may be worth it to get familiar with the code and IFC schema to make sure you extract the information you need.

**Comment by bingo-soft** [comment]
It is really strange, because in my IFC file all materials have id attribute. Probably, it is custom attribute

**Comment by Moult** [comment]
@bingo-soft you're probably thinking of the STEP ID. This is specific to the IFC-SPF format (.ifc file). An IfcMaterial itself has no ID. What viewer are you using which shows this ID attribute? Can you provide a screenshot?

**Comment by bingo-soft** [comment]
@Moult I see how IfcMaterialList is referenceing child IfcMaterials with indices right in text editor:
![photo_2024-07-30_16-38-12](https://github.com/user-attachments/assets/8e221405-ec0d-477e-bd72-1e1bd850e39a)

And I see this "id" attribute, when I parse ifc with  Xbim.Essentials library

**Comment by Moult** [comment]
Yeah those are STEP IDs, they are a detail of the SPF serialization and don't actually exist in the IFC schema. Don't rely on them as any sort of permanent key.
